### PR TITLE
Support Anima model

### PR DESCRIPTION
**Pull request by duongve13112002**
Hi @tdrussell
Here is the description of my change for support Anima model
1. Summary
- Add Anima model support (LoRA and full-finetune training)
- Add `max_checkpoints` config option to automatically delete old checkpoints

2. Changes

a. Anima Model Support
- Added `models/anima.py` - Anima pipeline with dual text encoders (Qwen3 + T5)
- Added `models/anima_modeling.py` - MiniTrainDIT with LLMAdapter architecture
- Support for both LoRA and full-finetune training modes
- Options: `freeze_llm_adapter`, `llm_adapter_lr`, `load_to_cuda`
- Compatible with Qwen NF4 quantization for lower VRAM

b.  Checkpoint Management
- Added `max_checkpoints` option in `utils/saver.py`
- Automatically deletes oldest checkpoints when limit exceeded
- Applies to: `global_step*`, `epoch*`, `step*` directories

Note: I checked for training LORA and full finetune model and it work correctly. If you have any question feel free to ask.

### CODE REVIEW

**Comment by frzfrzz** [comment]
hope this will be merged soon.

**Comment by urlesistiana** [comment]
Maybe something is off about embedding and masking. Padding tokens are not masked out. Model collapsed if no caption/condition.

File: models/anima.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        target_device = x_B_C_T_H_W.device
+        if qwen_embeds.device != target_device:
+            qwen_embeds = qwen_embeds.to(target_device)
+        if t5_input_ids.device != target_device:
+            t5_input_ids = t5_input_ids.to(target_device)
+        if t5_input_ids.dtype != torch.long:
+            t5_input_ids = t5_input_ids.long()
+
+        # Process through LLM adapter to get final cross-attention embeddings
+        crossattn_emb = self.llm_adapter(qwen_embeds, t5_input_ids)
```
Comment (urlesistiana) [inline models/anima.py:429]: https://github.com/tdrussell/diffusion-pipe/pull/505#issuecomment-3846131559

Also, do we need to zero out masked crossattn_emb as well? No training code for reference sucks...

Comment (duongve13112002) [reply models/anima.py:429]: @urlesistiana  i fixed, please check it out. However i am not sure about this according to comfyui they dont use attention mask, so i am not sure this is right or not

**Comment by tdrussell** [comment]
I appreciate the enthusiasm from the community to get training working. The official training code is in fact just a private branch of diffusion-pipe. I was going to delay training code for full release, since in all likelihood any loras trained on the preview version won't work well on the final version (the model still has a lot of training to do).

But I see that people have hacked in lora training elsewhere and are already pushing some loras on Civit. I'll go ahead and get the official code cleaned up and added ASAP.

**Comment by bluvoll** [comment]
> I appreciate the enthusiasm from the community to get training working. The official training code is in fact just a private branch of diffusion-pipe. I was going to delay training code for full release, since in all likelihood any loras trained on the preview version won't work well on the final version (the model still has a lot of training to do).
> 
> But I see that people have hacked in lora training elsewhere and are already pushing some loras on Civit. I'll go ahead and get the official code cleaned up and added ASAP.

Sorry for hacking training before you released the code :doro:

**Comment by motimalu** [comment]
Apologies for pushing loras a bit prematurely also, just very impressed with this model and its capabilities. It's amazing. 
I've included the lora disclaimer as noted in your documentation, and eagerly await the final model!

**Comment by tdrussell** [comment]
No worries, it is good that people are excited about the model. I just wanted to make sure that lora trainers know that their preview loras may fail in mysterious ways on the final version. Or maybe they work fine, who knows.

The official training code is now in and it should be identical in functionality to this PR so I think you can close it. Unless I messed something up cleaning up and copying over all the code which is possible.
